### PR TITLE
709

### PR DESCRIPTION
**Pull request by bserdar**

### CODE REVIEW

File: crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java [hunk 1]
```diff
@@ -54,6 +65,202 @@
...
+            }
+            throw new RuntimeException("Query processing error: cannot find clause in query field info:"+clause.toString());
+        }
+
+        private QueryFieldInfo[] findTwo(QueryExpression clause) {
+            QueryFieldInfo[] ret=new QueryFieldInfo[2];
+            int i=0;
+            for (QueryFieldInfo fi : fieldInfo) {
+                if (fi.getClause() == clause) {
+                    ret[i++]=fi;
```
Comment (BVulaj) [inline crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java]: Could this return more than two?

File: crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java [hunk 2]
```diff
@@ -46,6 +48,15 @@
     private static final long serialVersionUID = 1l;
 
     /**
+     * Type of the conjunct. Value means the conjunct compares a field
+     * to a value. relation means conjunct compares fields of an
+     * entity to the fields of another entity. Comples means either
```
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java]: Comples - typo

**Comment by bserdar** [comment]
fixed.

**Comment by bserdar** [comment]
Not in the way it is used. It is only called for leaf expressions, and
those can have either one or two entities in them.

File: crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java [hunk 3]
```diff
@@ -54,6 +65,202 @@
...
+                    first=false;
+                } else if(ret.type!=ConjunctType.complex) {                    
+                    ItrState qstate=iterate(subq,context);
+                    // We compare if this expression is compatible with existing assessment
+                    switch(qstate.type) {
+                    case value:
+                        if(ret.type==ConjunctType.value) {
+                            // Comparing value expression to another value expression
+                            // At most one entity can be present
+                            if(ret.numEntities==0||ret.e1==qstate.e1) {
```
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java]: If nothing happens, why not just negate the condition and move the else block up?

File: crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java [hunk 4]
```diff
@@ -54,6 +65,202 @@
...
+                                ; // No changes, compatible expressions
```
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java]: Semi-colon by itself if unnecessary.

File: crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java [hunk 5]
```diff
@@ -54,6 +65,202 @@
...
+            boolean first=true;
+            for(QueryExpression subq:q.getQueries()) {
+                if(first) {
+                    ret=iterate(subq,context);
+                    first=false;
+                } else if(ret.type!=ConjunctType.complex) {                    
+                    ItrState qstate=iterate(subq,context);
+                    // We compare if this expression is compatible with existing assessment
+                    switch(qstate.type) {
+                    case value:
```
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java]: There is a lot of logic in these case statements. Perhaps private methods might help the readability?

File: crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java [hunk 6]
```diff
@@ -88,6 +295,39 @@ public QueryExpression getClause() {
...
+               clause instanceof NaryFieldRelationalExpression) {
+                // These are simple 2-term expressions
+                return ConjunctType.relation;
+            } else {
+                // At this point, clause is one of:
+                //   - ArrayMatchExpression
+                //   - NaryLogicalExpression
+                //   - UnaryLogicalExpression
+                ClauseIterator itr=new ClauseIterator();
+                return itr.iterate(clause).type;
```
Comment (dcrissman) [inline crud/src/main/java/com/redhat/lightblue/assoc/Conjunct.java]: What about the case where an unexpected Class is passed in. Should an unsupported exception be thrown?

**Comment by bserdar** [comment]
done

**Comment by bserdar** [comment]
Done

**Comment by bserdar** [comment]
There can't be an unexpected class here. Any addition to query processing
logic affects the base class for query iterator.

**Comment by BVulaj** [comment]
@dcrissman @bserdar nothing outstanding.  Merging!

**Pull request merged**